Paginator fixes take3

billschaller opened the pull request:
This PR fixes issues reported in #1347 and #1351.
- Ordering a query by a column from an association which uses SINGLE_TABLE inheritance no longer throws an error.
- Limiting a query which is ordered by fields from a joined -To-Many association now works as expected.

This patch converts LimitSubqueryOutputWalker to use the ROW_NUMBER window function for queries on platforms that support it. Other platforms use a modified version of the previous method. The modification is to not select the ORDER BY columns in the wrapping query. The reason those columns were selected like that in the first place is that ORDER BY on columns not in the select list is unsupported on many platforms. MySQL and SQLite are fine with it, and don't support ROW_NUMBER.

The ROW_NUMBER solution is actually much simpler.

LimitSubqueryOutputWalker SQL generation tests for PG and Oracle have been changed.

**There are 2 failing expectedException tests right now.** The LimitSubqueryWalker needs to be modified to throw an exception when it detects a query that does the following things together:
- Joins a -To-Many association
- Orders by a column from the associated entity
- Limits the result

As with the previous patch, this depends on changes to SQLServerPlatform2008 that have not yet been merged. SQLServerPlatform2008 in DBAL/master currently mangles the generated queries when trying to apply LIMIT/OFFSET.

Commit:
problem with LimitSubqueryOutputWalker when use InheritanceType

Conflicts:
	tests/Doctrine/Tests/OrmFunctionalTestCase.php

Commit:
Fix LimitSubqueryOutputWalker when used on tables with joined table inheritance

Commit:
Fix paginator issues when ordering by a joined column from a to-many association.

Manual merge testcase from #1351

Commit:
Merge remote-tracking branch 'remotes/origin/DDC-3637-fix' into DDC-3606-testcase

Commit:
Cleanup

Commit:
Fix LimitSubqueryOutputWalkerTest

zeroedin-bill commented:
@vaheshadunts @alexander-orabey I believe this fixes both of your issues... Please test.

zeroedin-bill commented:
@mrkrstphr Does this fix your issue?

mrkrstphr commented:
@zeroedin-bill no, it does not. I will try to write a failing test tomorrow.

zeroedin-bill commented:
@mrkrstphr Ok, thanks - the functional testsuite is taking shape, and that should help prevent regressions like this in the future.

What platform are you running on?

mrkrstphr commented:
Test suite runs in sqlite.

vaheshadunts commented:
@zeroedin-bill as I see its ok in travis, I'll merge and test it later, thanks.

antanas-arvasevicius commented:
Hi, I don't know is it related but now this query is throwing exception
```
        $q = $this->entityManager->createQuery(
            '
SELECT entity FROM System\Entity\SimpleEntity entity
WHERE  ((SELECT COUNT(simple.id) FROM System\Entity\SimpleEntity AS simple) = 1)
ORDER BY entity.id DESC
        '
        );
        $q->setHint(Query::HINT_CUSTOM_OUTPUT_WALKER, 'Doctrine\ORM\Tools\Pagination\LimitSubqueryOutputWalker');
        $list = $q->getScalarResult();
```

Exception:
````
An exception occurred while executing 'SELECT DISTINCT id_0 FROM (SELECT s0_.id AS id_0, s0_.name AS name_1, s1_.id AS id_2 FROM simple_entity s0_ WHERE ((SELECT COUNT(s1_.id) AS dctrn__1 FROM simple_entity s1_) = 1)) dctrn_result ORDER BY id_0 DESC':

SQLSTATE[42S22]: Column not found: 1054 Unknown column 's1_.id' in 'field list'
```

If I remove `ORDER BY entity.id DESC` from query then it runs fine. 

Now its impossible to use Paginator when where is ordering and subquery with count()

mrkrstphr commented:
@zeroedin-bill it looks like our issue is cases that use table inheritance with an inheritance type of `JOINED`. 

For a test case, should I put that in `PaginatorTest` or in `Functional\Ticket` somewhere?

zeroedin-bill commented:
Since most of PaginationTest is new, and it hasn't been merged, please put
it into PaginationTest - I'll merge it into my PR branch.

Thanks!
On Mar 30, 2015 8:52 AM, "Kristopher Wilson" <notifications@github.com>
wrote:

> @zeroedin-bill <https://github.com/zeroedin-bill> it looks like our issue
> is cases that use table inheritance with an inheritance type of JOINED.
>
> For a test case, should I put that in PaginatorTest or in
> Functional\Ticket somewhere?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/1353#issuecomment-87665933>.
>

mrkrstphr committed:
test case for joined table inheritance pagination

Commit:
Add test for paginating on a query with a subquery in the where clause

Conflicts:
	tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryOutputWalkerTest.php

Commit:
Fix issue where paginating on a query with a subquery in the where clause crashed

zeroedin-bill commented:
@antanas-arvasevicius - fixed and added a testcase, thanks

mrkrstphr commented:
@zeroedin-bill here is a failing test case: https://github.com/doctrine/doctrine2/compare/master...mrkrstphr:paginator-test-for-joined-inheritance?expand=1

zeroedin-bill commented:
Great, thanks :)

Bill Schaller
Software and Support Engineer
Zeroed-In Technologies LLC
m:  +1.336.601.0632   o:  +1.410.242.6611 x2122
w:  *http://www.zeroedin.com <http://www.zeroedin.com/> *

On Mon, Mar 30, 2015 at 11:19 AM, Kristopher Wilson <
notifications@github.com> wrote:

> @zeroedin-bill <https://github.com/zeroedin-bill> here is a failing test
> case:
> https://github.com/doctrine/doctrine2/compare/master...mrkrstphr:paginator-test-for-joined-inheritance?expand=1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/1353#issuecomment-87720359>.
>

Commit:
Merge remote-tracking branch 'mrkrstphr/paginator-test-for-joined-inheritance' into paginator-fixes-take3

Conflicts:
	tests/Doctrine/Tests/ORM/Functional/PaginationTest.php

Commit:
Fix paginator when ordering by while selecting entities using joined table inheritance

zeroedin-bill commented:
@mrkrstphr - merged and fixed.

stof commented on lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php:
```diff
@@ -39,17 +49,17 @@ class LimitSubqueryOutputWalker extends SqlWalker
      * @var \Doctrine\DBAL\Platforms\AbstractPlatform
      */
-    private $platform;
+    private $_platform;
```
please don't add the underscore prefix. We are getting rid of it, not adding new ones

zeroedin-bill replied:
@stof - I actually ran into some really strange behavior with this. These properties were hiding the private ones from the parent class, but when I overrode walkSelectStatement, I was getting a fatal error saying that these private variables were undefined. It was really odd. I'll try backing the underscores out and see if I can reproduce.

Ideally these variables should just be protected on SqlWalker, but I didn't want to push changes too far out from LimitSubqueryOutputWalker.

zeroedin-bill replied:
Of course, when I try to reproduce, the error is gone...

zeroedin-bill replied:
@stof Please see http://www.doctrine-project.org/jira/browse/DDC-3688 - this is the issue I was talking about... Any ideas?

stof commented on lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php:
```diff
@@ -364,22 +450,106 @@ private function rebuildOrderByClauseForOuterScope(OrderByClause $orderByClause)
+     * @throws \Doctrine\ORM\OptimisticLockException
+     * @throws \Doctrine\ORM\Query\QueryException
+     */
+    private function getInnerSQL($AST)
```
should be typehinted

stof commented on lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php:
```diff
@@ -364,22 +450,106 @@ private function rebuildOrderByClauseForOuterScope(OrderByClause $orderByClause)
+     * @param $AST
+     * @return array
+     */
+    private function getSQLIdentifier($AST)
```
should be typehinted

stof commented on lib/Doctrine/ORM/Tools/Pagination/RowNumberOverFunction.php:
```diff
@@ -0,0 +1,35 @@
+<?php
+/**
+ * RowNumberOverFunction.php
```
This is the wrong file header

zeroedin-bill replied:
Is there a right header? I'm looking around at file headers and seeing some wide variation...

An example would be LimitSubqueryWalker.php:

```php
/**
 * Doctrine ORM
 *
 * LICENSE
 *
 * This source file is subject to the new BSD license that is bundled
 * with this package in the file LICENSE. This license can also be viewed
 * at http://hobodave.com/license.txt
 *
 * @category    DoctrineExtensions
 * @package     DoctrineExtensions\Paginate
 * @author      David Abdemoulaie <dave@hobodave.com>
 * @copyright   Copyright (c) 2010 David Abdemoulaie (http://hobodave.com/)
 * @license     http://hobodave.com/license.txt New BSD License
 */
```

CountOutputWalker.php:
```php
/**
 * Doctrine ORM
 *
 * LICENSE
 *
 * This source file is subject to the new BSD license that is bundled
 * with this package in the file LICENSE.txt.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to kontakt@beberlei.de so I can send you a copy immediately.
 */
```

AttachEntityListenersListener.php:
```php
/*
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software consists of voluntary contributions made by many individuals
 * and is licensed under the MIT license. For more information, see
 * <http://www.doctrine-project.org>.
 */
```

I assume this last one is correct, as it is the most recent incarnation?

zeroedin-bill replied:
@Ocramius proper file header where?

guilhermeblanco replied:
The last one... we're MIT, not BSD

zeroedin-bill replied:
Thanks :)

stof commented on lib/Doctrine/ORM/Tools/Pagination/RowNumberOverFunction.php:
```diff
@@ -0,0 +1,35 @@
+     * @override
+     */
+    public function parse(\Doctrine\ORM\Query\Parser $parser)
+    {}
```
You should throw an exception there, saying it cannot be used this way.

Commit:
resolve review comments from @stof

zeroedin-bill commented:
@stof resolved comments, thanks

antanas-arvasevicius commented:
@zeroedin-bill thank you very much!

Commit:
Add more detection in LimitSubqueryWalker for conditions that must be handled by LimitSubqueryOutputWalker

mrkrstphr committed:
Adding failing test case for PaginationTest

The failure comes into play when an entity has an attribute named
differently from its corresponding column name.

mrkrstphr commented:
@zeroedin-bill found another issue: https://github.com/zeroedin-bill/doctrine2/pull/1

mrkrstphr commented:
@zeroedin-bill commenting out this function call (https://github.com/zeroedin-bill/doctrine2/blob/paginator-fixes-take3/lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php#L424) seems to fix it, and doesn't cause any other errors. Is this call necessary?

Commit:
Merge pull request #1 from mrkrstphr/different-names

merged #1

Commit:
Fix issue in LimitSubqueryOutputWalker with entities having field names different from column names

zeroedin-bill commented:
@mrkrstphr fixed, thanks :)

mrkrstphr commented:
Our test suite now passes with this branch. Thanks!

zeroedin-bill commented:
@mrkrstphr Thanks for your help and testing :+1:

Ocramius commented:
Merged into `master` at doctrine/doctrine2@82230cc42706741699cd9bbfa4b59a7588ddcd1d - now tagging 2.5.0-RC2
